File: framework/src/run-support/src/main/scala/play/runsupport/AssetsClassLoader.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.runsupport

import java.io.File

/**
 * A ClassLoader for serving assets.
 *
 * Serves assets from the given directories, at the given prefix.
 *
 * @param assets A list of assets directories, paired with the prefix they should be served from.
 */
class AssetsClassLoader(parent: ClassLoader, assets: Seq[(String, File)]) extends ClassLoader(parent) {
  override def findResource(name: String) = {
    assets.collectFirst {
      case (prefix, dir) if exists(name, prefix, dir) =>
        new File(dir, name.substring(prefix.length)).toURI.toURL
    }.orNull
  }

  def exists(name: String, prefix: String, dir: File) = {
    name.startsWith(prefix) && new File(dir, name.substring(prefix.length)).isFile
  }
}

File: dev-mode/play-run-support/src/main/java/play/runsupport/classloader/AssetsClassLoader.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.runsupport.classloader;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.Map.Entry;

/**
 * A ClassLoader for serving assets. Serves assets from the given directories, at the given prefix.
 */
public class AssetsClassLoader extends ClassLoader {

  /** An assets directories, mapped by the prefix they should be served from. */
  private final List<Entry<String, File>> assets;

  public AssetsClassLoader(ClassLoader parent, List<Entry<String, File>> assets) {
    super(parent);
    this.assets = assets;
  }

  @Override
  public URL findResource(String name) {
    return assets.stream()
        .filter(e -> exists(name, e.getKey(), e.getValue()))
        .findFirst()
        .map(
            e -> {
              try {
                return new File(e.getValue(), name.substring(e.getKey().length())).toURI().toURL();
              } catch (MalformedURLException ex) {
                return null;
              }
            })
        .orElse(null);
  }

  private boolean exists(String name, String prefix, File dir) {
    return name.startsWith(prefix) && new File(dir, name.substring(prefix.length())).isFile();
  }
}
